cluster-autoscaler/proposals/provisioning-request.md

```diff
@@ -0,0 +1,293 @@
+	//
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Spec contains specification of the ProvisioningRequest object.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status.
+	//
+	// +kubebuilder:validation:Required
+	Spec ProvisioningRequestSpec `json:"spec" protobuf:"bytes,2,name=spec"`
+	// Status of the ProvisioningRequest. CA constantly reconciles this field.
+	//
+	// +optional
+	Status ProvisioningRequestStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// ProvisioningRequestList is a object for list of ProvisioningRequest.
+type ProvisioningRequestList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list metadata.
+	//
+	// +optional
+	metav1.ListMeta `json:"metadata" protobuf:"bytes,1,opt,name=metadata"`
+	// Items, list of ProvisioningRequest returned from API.
+	//
+	// +optional
+	Items []ProvisioningRequest `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// ProvisioningRequestSpec is a specification of additional pods for which we
+// would like to provision additional resources in the cluster.
+type ProvisioningRequestSpec struct {
+	// PodSets lists groups of pods for which we would like to provision
+	// resources.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Value is immutable"
+	PodSets []PodSet `json:"podSets" protobuf:"bytes,1,rep,name=podSets"`
+
+	// ProvisioningClass describes the different modes of provisioning the resources.
+	// Supported values:
+	// * GenericCheckCapacity - check if current cluster state can fullfil this request
+	// * GenericAtomicScaleUp - provision the resources in an atomic manner
+    // * ... - potential other classes that are specific to the cloud providers
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Value is immutable"
+	ProvisioningClass string `json:"provisioningClass" protobuf:"bytes,2,name=provisioningClass"`
+
+	// ValidUntilSeconds contains information on how long we can wait for the ProvReq to provide VMs.
+	// Field specific to GenericAtomicScaleUp provisioning class.
+	//
+	// +optional
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Value is immutable"
+	ValidUntilSeconds *int64 `json:"validUntilSeconds" protobuf:"bytes,3,name=validUntilSeconds"`
+
+	// AdditionalParameters contains all other parameters custom classes may require.
+	//
+	// +optional
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Value is immutable"
+	AdditionalParameters map[string]string `json:"additionalParameters" protobuf:"bytes,4,name=additionalParameters"`
+}
+
+type PodSet struct {
+	// Template representing pods that will consume this reservation.
+	// Requirements for resources (CPU, RAM, GPUs, TPUs, storage) are
+	// necessary (must be non-zero). Users need to make sure that the
```

Comment:
what about requirements / tolerations / selectors that are not known ahead of pod creation time, or are modified at pod creation time by admission plugins? the approach below where the pod spec info is pulled from actual pod or podtemplate instances could resolve this issue

Reply:
But this would also introduce other issues, what if each of the 1000 pods has a different set of tolerations and selectors, how should CA pick the representative?
This also does not solve the issue where the pods are pending and are unschedulable generating load for scheduler and noise in the alerts/logs.
And if we were to fetch the pod template from the workload definitions then CA would be required to understand each k8s object type that can result in creation of pod, which generates a lot of other problems.

Reply:
> what if each of the 1000 pods has a different set of tolerations and selectors, how should CA pick the representative?

if that happens, this design won't work properly, right? the individual pods will instruct CA to ignore them, and CA will provision capacity based on the single podspec in the provisioningrequest, right?

Reply:
> if that happens, this design won't work properly, right? the individual pods will instruct CA to ignore them, and CA will provision capacity based on the single podspec in the provisioningrequest, right?

Yes

> CA reading from Pod objects and PodTemplate objects covers this case. A single PodTemplate would communicate the spec to CA without requiring spawning pods.

But in this case users would need to create one template pod (or PodTemplate), which further complicates the user interaction with this feature.
We can add both options to the CRD and remove the embedded pod template once we move the CRD to v1.
WDYT?

Reply:
> But in this case users would need to create one template pod (or PodTemplate), which further complicates the user interaction with this feature.

Two objects is not always more complicated than one... especially if existing tooling / validation / admission works better when creating podtemplate or pod or workload objects. Are we expecting users to construct provisioning request manifests by hand? If so, the duplication leading to inaccurate copies is also a potential user experience hazard with the current design.

I think embedding podspec into many different CRDs is a locally optimized decision that ends up being a _worse_ user experience overall.

> We can add both options to the CRD and remove the embedded pod template once we move the CRD to v1.
WDYT?

If the embedded spec makes it into the CRD, I don't believe it will ever get dropped.

Reply:
Do you know what native k8s objects allow users to reference the PodTemplates?

Reply:
There is also an issue of snapshotting - capturing the exact spec that was used for provisioning. Some modes of provisioning won't allow for late changes, e.g. if queuing is involved. Inline spec is cleaner for that, as we can validate mutability and it's easy to check what was used for provisioning. If we refer to external objects such as PodTemplate or the workload, it's harder to track, but I guess we could store the snapshotted revision in ProvisioningRequest status.

> I was referring to the single PodTemplate API, not looking inside deployment/job/etc objects

PodTemplate doesn't really help with having a single source of truth if real deployment/job/pods don't use it, right? What's the benefit of the separate PodTemplate if it's only used by ProvisioningRequest?


For these reasons, I think we should choose between
1. Inline spec as currently proposed (better for snapshotting and CA clients/watches/caches, but with downsides of duplicating the spec)
2. Link to the pod template in the workload object such as Job -- gives a single source of truth, but more problematic in terms of snapshotting and CA clients/watches/caches.

Reply:
Another consideration (speaking for inline spec) is ability to provision with incomplete spec if the information is not needed for provisioning. It should be fine to provision nodes before knowing the exact container image.

Reply:
Regarding the ability to provision for incomplete specs: What are the required fields in a PodTemplate that users might not know beforehand? I cannot think of anything other than container image. If so, we could ask users to inject anything or we could do it through a webhook. Long term, we could relax the validation in apiserver to allow empty container image in PodTemplate.

Reply:
Users should be able to inject anything that is not taken into account by the default scheduler, the biggest one being the container images and names

Reply:
That's my point, container image and name are the only mandatory fields which could easily be mocked by a webhook.
As a stable API, it's almost impossible to add mandatory fields to Pod, so this the set of mandatory fields is unlikely to change.
@liggitt wdyt?

Reply:
> this the set of mandatory fields is unlikely to change.

that's true

Reply:
Moved to the PodTemplate, closing the thread for now